godot-core/src/builtin/rect2i.rs

```diff
@@ -78,21 +82,492 @@ impl Rect2i {
         self.size = end - self.position
     }
 
-    /* Add in when `Rect2i::abs()` is implemented.
+    /// Returns a `Rect2i` with equivalent position and area, modified so that the top-left corner
+    /// is the origin and `width` and `height` are positive.
+    #[inline]
+    pub fn abs(self) -> Self {
+        let abs_size = self.size.abs();
+        let offset = Vector2i::new(cmp::min(self.size.x, 0), cmp::min(self.size.y, 0));
+        Self::new(self.position + offset, abs_size)
+    }
+
+    /// Returns `true` if this `Rect2i` completely encloses another one.
+    #[inline]
+    pub const fn encloses(&self, other: Self) -> bool {
+        self.assert_nonnegative();
+        other.assert_nonnegative();
+        // FIXME: provide an unsafe version without the check?
+        // FIXME: or just remove the check altogether?
+        let own_end = self.end();
+        let other_end = other.end();
+        other.position.x >= self.position.x
+            && other.position.y >= self.position.y
+            && other_end.x <= own_end.x
+            && other_end.y <= own_end.y
+    }
+
+    /// Returns a copy of this `Rect2i` expanded so that the borders align with the given point.
+    #[inline]
+    pub fn expand(self, to: Vector2i) -> Self {
+        self.assert_nonnegative();
+        // FIXME: provide an unsafe version without the check?
+        // FIXME: or just remove the check altogether?
+        let mut end = self.end();
+        let mut begin = self.position;
+
+        begin.x = cmp::min(begin.x, to.x);
+        begin.y = cmp::min(begin.y, to.y);
+        end.x = cmp::max(end.x, to.x);
+        end.y = cmp::max(end.y, to.y);
+
+        Self::from_corners(begin, end)
+    }
+
+    /// Returns the area of the `Rect2i`.
+    #[inline]
+    pub const fn get_area(&self) -> i32 {
+        self.size.x * self.size.y
+    }
+
+    /// Returns the center of the `Rect2i`, which is equal to `position + (size / 2)`.
+    ///
+    /// If `size` is an odd number, the returned center value will be rounded towards `position`.
+    #[inline]
+    pub fn get_center(&self) -> Vector2i {
+        self.position + (self.size / 2)
+    }
+
+    /// Returns a copy of the `Rect2i` grown by the specified `amount` on all sides.
+    #[inline]
+    pub fn grow(self, amount: i32) -> Self {
+        let amount_2d = Vector2i::new(amount, amount);
+        Self::from_corners(self.position - amount_2d, self.end() + amount_2d)
+    }
```

Comment:
For all the `grow*` functions: what happens if `amount` is negative? Is it something we should protect against? 

Godot doesn't document a behavior, do you know what it's doing?

Reply:
This is the relevant Godot code: https://github.com/godotengine/godot/blob/master/core/math/rect2i.h#L154

It generally works, at least as long as the size does not become negative.

Negative grow is covered in some of the tests as well.

Reply:
> It generally works, at least as long as the size does not become negative.
>
> Negative grow is covered in some of the tests as well.

Could you document that behavior? Both that `grow <= 0` is allowed, and that too much negative grow can turn the rectangle's size negative? If the user expects "shrink but stop at 0" behavior, they can call `Vector2i::cwise_max(rect.size, Vector2i::ZERO)` on the result.

Also, this operation is allowed on negative rectangles in the first place? Would probably make sense, as the operation allows negative rectangles in the first place. That is, `rect.grow(i).grow(-i)` for any `i` would always return the original rectangle `rect`.